client/client-multi/client-c/src/kaa_channel_manager.c

```diff
@@ -14,135 +14,283 @@
-        , kaa_sync_handler_fn handler, const kaa_service_t *supported_services, size_t services_count)
+kaa_error_t kaa_channel_manager_add_transport_channel(kaa_channel_manager_t *self
+                                                    , kaa_transport_channel_interface_t *channel)
 {
-    KAA_RETURN_IF_NIL4(self, handler, supported_services, services_count, KAA_ERR_BADPARAM);
+    KAA_RETURN_IF_NIL2(self, channel, KAA_ERR_BADPARAM);
 
-    kaa_sync_details *sync = (kaa_sync_details *) KAA_MALLOC(sizeof(kaa_sync_details));
-    if (!sync)
-        return KAA_ERR_NOMEM;
+    kaa_transport_protocol_info_t protocol_info;
+    channel->get_protocol_info(channel->context, &protocol_info);
+
+    kaa_list_t *it = kaa_list_find_next(self->transport_channels, &find_channel, channel);
+    if (it) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_ALREADY_EXISTS,
+                "Failed to add transport channel (id=0x%X, version=%u): already exists"
+                                                 , protocol_info.id, protocol_info.version);
+        return KAA_ERR_ALREADY_EXISTS;
+    }
+
+    kaa_transport_channel_interface_t *copy =
+            (kaa_transport_channel_interface_t *)KAA_MALLOC(sizeof(kaa_transport_channel_interface_t));
+    KAA_RETURN_IF_NIL(copy, KAA_ERR_NOMEM);
+
+    *copy = *channel;
 
-    sync->sync_fn = handler;
-    sync->supported_services_size = services_count;
-    sync->supported_services = (kaa_service_t *) KAA_MALLOC(services_count * sizeof(kaa_service_t));
-    while (services_count--)
-        sync->supported_services[services_count] = supported_services[services_count];
-
-    // Add to the list of handlers (or create the list if there's none yet)
-    self->sync_handlers = self->sync_handlers
-            ? kaa_list_push_front(self->sync_handlers, sync)
-            : kaa_list_create(sync);
-    if (!self->sync_handlers) {
-        KAA_FREE(sync->supported_services);
-        KAA_FREE(sync);
+    it = self->transport_channels ?
+                 kaa_list_push_front(self->transport_channels, copy) :
+                 kaa_list_create(copy);
+    if (!it) {
+        KAA_LOG_ERROR(self->logger, KAA_ERR_NOMEM,
+                "Failed to add new transport channel (id=0x%X, version=%u)"
+                                                     , protocol_info.id, protocol_info.version);
+        KAA_FREE(copy);
         return KAA_ERR_NOMEM;
     }
 
+    self->transport_channels = it;
+    self->sync_info.is_up_to_date = false;
+
+    KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "New transport channel (id=0x%X, version=%u) added"
+                                                                        , protocol_info.id, protocol_info.version);
+
     return KAA_ERR_NONE;
 }
 
-kaa_error_t kaa_channel_manager_remove_sync_handler(kaa_channel_manager_t *self, kaa_sync_handler_fn handler)
+kaa_error_t kaa_channel_manager_remove_transport_channel(kaa_channel_manager_t *self
```

Comment:
I guess such "remove" API wouldn't work in practice. User has got only two ways to use it:
1) Store his "kaa_transport_channel_interface_t" instance in the heap or static memory.
2) User should create and fill "kaa_transport_channel_interface_t" on stack when he's adding a channel and then create and fill the same instance when he's trying to remove it.

Both ways are not very usable IMHO.